### PR TITLE
V3—Case plot and dot chart

### PR DESCRIPTION
**Pull request by bfinzer**
This pull requests has major effects on the user experience with graphs:
* The initial graph is a case plot with no attributes on either axis
* Movable value and movable lin no longer appear
* Dragging a categorical attribute to either axis results in a dot chart
* Dragging a numeric attribute to the vertical axis results in a dot plot that is a rotation of the one from the x-axis

There are also numerous bug fixes, especially having to do with graph and axis layout

### CODE REVIEW

File: v3/src/components/graph/utilities/graph_utils.ts [hunk 1]
```diff
@@ -74,31 +71,29 @@ export function computeNiceNumericBounds(min: number, max: number): { min: numbe
...
-  axis.setTransitionDuration(1000)
-  axis.setDomain(niceBounds.min, niceBounds.max)
-  scale?.domain([niceBounds.min, niceBounds.max])  // We can't rely on useNumericAxis hook because of the transition
+export function setNiceDomain(values: (string | number)[], axisModel: IAxisModel) {
+  if (axisModel.isNumeric) {
+    const numericAxisModel = axisModel as INumericAxisModel,
+      valueExtent = extent(values, d => Number(d)) as unknown as [number, number],
+      niceBounds = computeNiceNumericBounds(valueExtent[0], valueExtent[1])
+    numericAxisModel.setDomain(niceBounds.min, niceBounds.max)
+  }
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph_utils.ts:80]: Can be simplified to:
```suggestion
export function setNiceDomain(values: number[], axisModel: IAxisModel) {
  if (isNumericAxisModel(axisModel)) {
    const [minValue, maxValue] = extent(values, d => d) as [number, number],
      { min: niceMin, max: niceMax } = computeNiceNumericBounds(minValue, maxValue)
    axisModel.setDomain(niceMin, niceMax)
  }
```
making use of the type guard function that I added in a separate commit.

File: v3/src/components/graph/hooks/use-numeric-axis.test.tsx [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { renderHook } from "@testing-library/react"
 import React from "react"
 import { INumericAxisModel, NumericAxisModel } from "../models/axis-model"
 import { GraphLayout, GraphLayoutContext } from "../models/graph-layout"
-import { useNumericAxis } from "./use-numeric-axis"
+import { useAxis } from "./use-axis"
```
Comment (kswenson) [inline v3/src/components/graph/hooks/use-numeric-axis.test.tsx:6]: File should be renamed to `use-axis.test.ts` to match.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Overall this looks good -- I made a bunch of comments/suggestions but most of them are minor and/or discretionary. I also added a couple of commits to add a utility function, fix lint errors, and fix the cypress test.

File: v3/src/components/graph/components/axis.tsx [hunk 3]
```diff
@@ -44,12 +44,12 @@ export const Axis = ({attributeID, model, transform, onDropAttribute}: IProps) =
...
     const droppedAttrId = active.data?.current?.attributeId
-    droppedAttrId && onDropAttribute(model.place, droppedAttrId)
-  }, [model.place, onDropAttribute])
+    droppedAttrId && onDropAttribute(axisModel.place, droppedAttrId)
+  }, [axisModel.place, onDropAttribute])
 
   const data: IDropData = {accepts: ["attribute"], onDrop: handleDrop}
 
-  const [xMin, xMax] = scale.range()
+  const [xMin, xMax] = (scale as ScaleNumericBaseType).range()
```
[the remark this answers is not in the document]
Comment (bfinzer) [reply v3/src/components/graph/components/axis.tsx:52]: Type assertion not needed. All scales have a range function.

File: v3/src/components/graph/components/axis.tsx [hunk 4]
```diff
@@ -103,8 +103,8 @@ export const Axis = ({attributeID, model, transform, onDropAttribute}: IProps) =
         <g className='axis' ref={elt => setAxisElt(elt)}/>
         <g ref={titleRef}/>
       </g>
-      <AxisDragRects axisModel={model} axisWrapperElt={wrapperElt}/>
-      <DroppableAxis place={`${model.place}`} dropId={droppableId} dropData={data}
+      <AxisDragRects axisModel={axisModel as INumericAxisModel} axisWrapperElt={wrapperElt}/>
```
[the remark this answers is not in the document]
Comment (bfinzer) [reply v3/src/components/graph/components/axis.tsx:106]: Good catch!

File: v3/src/components/graph/components/chartdots.tsx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    values?.forEach(aValue => {
+      if (bins[aValue] === undefined) {
+        bins[aValue] = 0
+      }
+      bins[aValue]++
+    })
+    return Number((Object.keys(bins) as Array<keyof typeof bins>).reduce((max: number, aKey) => {
+      max = Math.max(max, bins[aKey])
+      return max
+    }, 0))
```
[the remark this answers is not in the document]
Comment (bfinzer) [reply v3/src/components/graph/components/chartdots.tsx:45]: Nice!

File: v3/src/components/graph/models/axis-model.ts [hunk 6]
```diff
@@ -16,31 +20,43 @@ export const AxisModel = types.model("AxisModel", {
...
+    scale: "ordinal"
   })
 export interface ICategoricalAxisModel extends Instance<typeof CategoricalAxisModel> {}
 
 export const NumericAxisModel = AxisModel
   .named("NumericAxisModel")
   .props({
     type: "numeric",
-    scale: types.optional(types.enumeration([...ScaleTypes]), "linear"),
+    scale: 'linear',
```
[the remark this answers is not in the document]
Comment (bfinzer) [reply v3/src/components/graph/models/axis-model.ts:59]: Not sure how that got changed, but I put that possibility back

File: v3/src/components/graph/components/casedots.tsx [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import {useDataSetContext} from "../../../hooks/use-data-set-context"
+import {useInstanceIdContext} from "../../../hooks/use-instance-id-context"
+import {ScaleNumericBaseType, useGraphLayoutContext} from "../models/graph-layout"
+import {setPointSelection} from "../utilities/graph_utils"
+
+export const CaseDots = memo(function CaseDots(props: {
+  dotsRef: React.RefObject<SVGSVGElement>
+  enableAnimation: React.MutableRefObject<boolean>
+}) {
+  useInstanceIdContext()
```
Comment (kswenson) [inline v3/src/components/graph/components/casedots.tsx:14]: Still appears unused/unnecessary.

**Pull request merged**